src/backend/distributed/deparser/deparse_extension_stmts.c

```diff
@@ -86,54 +89,72 @@ DeparseCreateExtensionStmt(Node *node)
-
-	const char *extensionName = createExtensionStmt->extname;
-	extensionName = quote_identifier(extensionName);
+	appendStringInfoString(buf, "CREATE EXTENSION ");
 
-	/*
-	 * We fetch "new_version", "schema" and "cascade" options from
-	 * optionList as we will append "IF NOT EXISTS" clause regardless of
-	 * statement's content before propagating it to worker nodes.
-	 * We also do not care old_version for now.
-	 */
-	Value *schemaNameValue = GetExtensionOption(optionsList, "schema");
-
-	/* these can be NULL hence check before fetching the stored value */
-	Value *newVersionValue = GetExtensionOption(optionsList, "new_version");
-	Value *cascadeValue = GetExtensionOption(optionsList, "cascade");
+	if (createExtensionStmt->if_not_exists)
+	{
+		appendStringInfoString(buf, "IF NOT EXISTS ");
+	}
 
 	/*
-	 * We do not check for if schemaName is NULL as we append it in deparse
-	 * logic if it is not specified.
+	 * Up until here we have been ending the statement in a space, which makes it possible
+	 * to just append the quoted extname. From here onwards we will not have the string
+	 * ending in a space so appends should begin with a whitespace.
 	 */
-	const char *schemaName = strVal(schemaNameValue);
-	schemaName = quote_identifier(schemaName);
+	appendStringInfoString(buf, quote_identifier(createExtensionStmt->extname));
+	AppendCreateExtensionStmtOptions(buf, createExtensionStmt->options);
+	appendStringInfoString(buf, ";");
+}
 
-	appendStringInfo(buf, "CREATE EXTENSION IF NOT EXISTS %s WITH SCHEMA %s",
-					 extensionName, schemaName);
 
-	/* "new_version" may not be specified in CreateExtensionStmt */
-	if (newVersionValue)
+/*
+ * AppendCreateExtensionStmtOptions takes the option list of a CreateExtensionStmt and
+ * loops over the options to add them to the statement we are building.
+ *
+ * An error will be thrown if we run into an unsupported option, comparable to how
+ * postgres gives an error when parsing this list.
+ */
+static void
+AppendCreateExtensionStmtOptions(StringInfo buf, List *options)
+{
+	if (list_length(options) > 0)
 	{
-		const char *newVersion = strVal(newVersionValue);
-		newVersion = quote_identifier(newVersion);
-
-		appendStringInfo(buf, " VERSION %s", newVersion);
+		/* only append WITH if we actual will add options to the statement */
+		appendStringInfoString(buf, " WITH");
 	}
 
-	/* "cascade" may not be specified in CreateExtensionStmt */
-	if (cascadeValue)
+	/* Add the options to the statement */
+	DefElem *defElem = NULL;
+	foreach_ptr(defElem, options)
 	{
-		bool cascade = intVal(cascadeValue);
-
-		if (cascade)
+		if (strcmp(defElem->defname, "schema") == 0)
+		{
+			const char *schemaName = defGetString(defElem);
+			appendStringInfo(buf, " SCHEMA  %s", quote_identifier(schemaName));
+		}
+		else if (strcmp(defElem->defname, "new_version") == 0)
 		{
-			appendStringInfoString(buf, " CASCADE");
+			const char *newVersion = defGetString(defElem);
```

Comment:
Optional: Implementing those function, I wasn't aware of such `defGet<type>` functions. Can we change `GetExtensionOption` to a function returning the `defElem` itself and change those following `strVal`, `intVal` etc. calls to `defGet<type>` across `extension.c` & `deparse_extension_stmts.c` ?

Reply:
Looks like a good improvement to me, I will add a commit on top doing this.